ex03.exs

```diff
@@ -1,110 +1,135 @@
+  process all elements of the original collection in parallel. If we
+  had one processor for each element in the original collection, that
+  would turn it into an O(1) operation in terms of elapsed time.
+
+  However, we don't have that many processors on our machines, so we
+  have to compromise. If we have two processors, we could divide the
+  map into two chunks, process each independently on its own
+  processor, then combine the results.
+
+  You might think this would halve the elapsed time, but the reality
+  is that the initial chunking of the collection and the eventual
+  combining of the results both take time. As a result, the speed up
+  will be less that a factor of two. If the work done in the mapping
+  function is time consuming, then the speedup factor will be greater,
+  as the overhead of chunking and combining will be relatively less.
+  If the mapping function is trivial, then parallelizing the code will
+  actually slow it down.
+
+  Your mission is to implement a function
+
+      pmap(collection, process_count, func)
+
+  This will take the collection, split it into n chunks, where n is
+  the process count, and then run each chunk through a regular map
+  function, but with each map running in a separate process. It then
+  combines the results (in the correct order). It should use
+  spawn and message passing (and not agents, tasks, or genservers).
+  It should not use any conditional logic (if/cond/case).
+
+  Useful functions include `Enum.map/3`, `Enum.chunk_every/4`, and
+  `Enum.flat_map/1`.
+
+  Feel free to use one or more helper functions... (there may be some
+  extra credit for code that is well factored and that looks good).
+  My solution is about 40 lines (including some blank ones) and
+  six helper functions.
+
+  35 points:
+     it works and passes all tests:    25
+     it contains no conditional logic:  3
+     it is nicely structured            7
+  """
+
+  def pmap(collection, process_count, function) do
+    divideCollection(collection, process_count) 
+    #divide the collection into chunks
+    |> mappingTogether(function) 
+    #run each chunk and mapping it
+    |> List.flatten
+    #collect chunks and merge them into a flat list
+  end
+
+  def divideCollection(collection, process_count) do
+    Enum.chunk_every(collection,div(Enum.count(collection),process_count))
+    #chunk_every(enum,cout)
+    #divide collection into required number of parts 
+  end
+
+  #mapping each chunk
+  def mappingTogether(chunks, function) do
+    thisOne = self()
+    #I dont know why, but if I use self() in following function, the program run into infinite loop
+    #using thisOne to avoid troubles if there is any build-in variable called "this"
+    Enum.map(chunks,fn chunk -> spawn (fn -> mapSubchunk(function,chunk,thisOne)end) end)
+    |>Enum.map(fn pid -> (receive do {^pid, value} -> value end) end)
+    #for each chunk divided in divideCollection, mapping each value
+  end

```

Comment:
This function does two different things (spawning the processes and waiting for the results). I'd much prefer to see this split into at least two functions.

# -2